jpsonic-main/src/main/java/com/tesshu/jpsonic/dao/MediaFileDao.java

```diff
@@ -140,6 +144,37 @@ public List<MediaFile> getChildrenWithOrderOf(String path) {
                 """, rowMapper, path);
     }
 
+    public List<MediaFile> getDirectChildren(MusicIndex musicIndex, List<MusicFolder> folders, long offset,
+            long count) {
+        Map<String, Object> args = LegacyMap.of("directory", MediaFile.MediaType.DIRECTORY.name(), "album",
+                MediaFile.MediaType.ALBUM.name(), "music", MediaFile.MediaType.MUSIC.name(), "audiobook",
+                MediaFile.MediaType.AUDIOBOOK.name(), "video", MediaFile.MediaType.VIDEO.name(), "musicIndex",
+                musicIndex.getIndex(), "folders", MusicFolder.toPathList(folders), "count", count, "offset", offset);
+        return template.namedQuery("select " + QUERY_COLUMNS + """
+                        ,
+                        case type
+                            when :directory then 1
+                            when :album then 2
+                            when :music then 3
+                            when :audiobook then 4
+                            when :video then 5
+                        end as type_order
+                from media_file
+                where music_index=:musicIndex and present and folder in (:folders)
+                order by type_order, media_file_order
+                offset :offset limit :count
+                """, rowMapper, args);
+    }
+
+    public List<MediaFile> getArtists(MusicIndex musicIndex, List<MusicFolder> folders, long offset, long count) {
+        return template.query("select " + QUERY_COLUMNS + """
+                from artist
+                where music_index=?
+                order by artist_order
+                offset ? limit ?
+                """, rowMapper, musicIndex.getIndex(), offset, count);
+    }
```

Comment:
In this method, only directory, album, and music can be included in the result. This should be redesigned...　Instead of rewriting the SQL, you may be able to share it with other logic by adding it to the filter condition in the method Paramater.

Well, in the case of a web page, for example, if you mix video and music in the same directory, I think only the music will be displayed. (Although it's a little vague) If there was only a video, it probably would have jumped to a dedicated View. 

Browsing must be designed so that there are basically no omissions. However, it is standard for media-related software to generate different tree nodes for each media type. (Otherwise, when you add a directory to the queue, music and videos will be mixed together in the queue.)

Reply:
#### Methods to be fixed

- MediaFileUpnpProcessor#getChildren
- MediaFileUpnpProcessor#getChildSizeOf
- MediaFileDao#getSingleSongs
- IndexUpnpProcessor#getChildren
- IndexUpnpProcessor#getChildSizeOf

#### Problem

We need to filter the Mediatype accurately when expanding the directory. For the above method:

 - Permissible : MediaType.MUSIC, MediaType.DIRECTORY, MediaType.ALBUM
 - Not acceptable : MediaType.PODCAST, MediaType.AUDIOBOOK, MediaType.VIDEO

This is because it is a tree node for handling music. Users browse the directory and often play the entire directory. At that time, media other than music must not be mixed. PODCAST, AUDIOBOOK, VIDEO should each be provided with a different tree node. (PODCAST is already implemented. AUDIOBOOK, VIDEO will also be created in the future) These are not specifications limited to Jpsonic in particular. Many UPnP media servers provide such browsing for user convenience.

#### Approach

A new method will be created that takes the MediaType filter as an argument, taking into account offset/count/order like the existing method. It is currently necessary, and is expected to be necessary at least for browsing AUDIOBOOK and VIDEO on UPnP.

We are currently considering web pages. (The design is different from UPnP) It won't be addressed right away. But personally I feel that it would be easier to understand if AUDIOBOOK and VIDEO browsing were isolated like UPnP. The view design of Subsonic is mixed type.

Reply:
Done.